File: 1612-avoid-flood-in-the-city/avoid-flood-in-the-city.cpp
class Solution {
public:
    vector<int> avoidFlood(vector<int>& rains) {
        map<int,int>mp;
        set<int>zeros;
        int n = rains.size();
        vector<int>ans(n,1);
        for(int i = 0 ; i < n;i++){
            if(!rains[i]){
                zeros.insert(i);
                continue;
            }
            if(mp.find(rains[i]) != mp.end()){
                auto it = zeros.lower_bound(mp[rains[i]]);
                if(it == zeros.end())
                    return {};
                ans[*it] = rains[i];
                zeros.erase(it);
                
            }
            mp[rains[i]] = i;
            ans[i]= -1;
        }
        return ans;

    }
};

File: 1612-avoid-flood-in-the-city/avoid-flood-in-the-city.py
class Solution:
    def avoidFlood(self, rains: List[int]) -> List[int]:
        zeros = SortedList()
        mp = {}
        n = len(rains)
        ans = [1]*n
        for i in range(n):
            if rains[i] == 0:
                zeros.add(i)
                continue
            if rains[i] in mp:
                it = zeros.bisect(mp[rains[i]])
                if it == len(zeros):
                    return []
                ans[zeros[it]] = rains[i]
                zeros.discard(zeros[it])
            mp[rains[i]] = i
            ans[i] = -1
        return ans

